### DISCUSSION TITLE
Looking for a way to unit test custom (validation) types

### DISCUSSION THREAD

**Discussion by lagnat** [mikro-orm/mikro-orm#3708] [General]
On the surface this looks like a simple question however my custom types are used for validation and can be parameterized so simple unit testing the custom type classes is not enough.  For example,

```
    @Property({
        fieldName: 'name',
        type: new MyStringType({
            not: /^-|-$|^[0-9]+$/,
            is: /^[0-9a-z-]+$/,
            len: [1, 63],
            truncate: 128,
        }),
    })
```

To unit test this, what I'd like to be able to do is take some known bad and known good data and execute some code that exercises custom type conversion.  I was able to perform the first half with this:

```
const raw = qb().helper.processData(attrs, true);
```

Next I'd like to be able to convert it back into an entity, so that I can exercise the other half of the custom types, but I wasn't able to find a way to do that.  Some portions of EntityManager map and EntityFactory create look useful but I don't want the merge.

The example above is a good use-case because of the `truncate` option.  I want to be able to unit test that this entity will truncate a long name so I need to be able to set a long name in an entity instance, and simulate what will happen inside of mikro and then be able to get the name back out and see that it was truncated.

**Comment by B4nan** [comment]
You want to unit test the type, then do so directly with the type class. Don't test the ORM when you want to test your own code.

**Comment by lagnat** [comment]
I tried to short circuit that answer but I accept it if that's the answer.  What I'm trying to test is the entity that has a specific configuration for a custom type.  Each property can use the same custom type but be configured differently.  My business rules say that the `name` field must be truncated over 128 chars.  I would like a test for that specific business rule. Create an entity, set a really long name, exercise the ORM, and check that the name is truncated.  I can test it with an actual database but prefer not to if I can get away with it.
